File: 3871_Count_Commas_in_Range_II/sol.ts
/**
 * Time:   O(log N)
 * Space:  O(1)
 * Strategy: Maths
 */
const solve = (n: number): number => {
    let result = 0;
    for (let i = 1000; i <= n; i*=1000) {
        result += n-i +1;
    }
    return result;
};
/**
 * Local testing
 */
interface TestCase {
    args: number; 
    expected: number;
}
function test(): void {
    const testCases: TestCase[] = [
        { args: 1002, expected: 3 },
        { args: 998, expected: 0 },
        { args: 1004590, expected: 1008182 }
    ];
    testCases.forEach(({ args, expected }, index) => {
        const result = solve(args);
        const isPassed = result === expected;
        const status = isPassed ? "✅ PASS" : "❌ FAIL";
        console.log(
            `${status} | Test #${index + 1}: ` +
            `Input: ${JSON.stringify(args)} | ` +
            `Expected: ${expected} | ` +
            `Result: ${result}`
        );
    });
}
test();
export {}; // This forces the file to be treated as a module with its own scope

File: 3871_Count_Commas_in_Range_II/sol.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    // Strategy: Maths
    // Time:     O(log N)
    // Space:    O(1)

    long long solve(const long long n) {
        long long result = 0;
        for (size_t i = 1000; i <= n; i *=1000){
            result += n-i +1;
        }return result;
    }
};
struct TestCase {
    long long input1;
    long long expected;
};
void run_tests() {
    Solution sol;
    vector<TestCase> testCases{
        {1002, 3}, 
        {998, 0},
        {1004590, 1008182}
    };
    for(size_t i = 0; i < testCases.size(); ++i) {
        int result = sol.solve(testCases[i].input1);
        bool isCorrect = (result == testCases[i].expected);
        cout << (isCorrect ? "✅" : "❌") << " Test #" << i;
        cout << " | Expected: " << testCases[i].expected 
             << " | Result: " << result << endl;
    }
}
int main() {
    run_tests();
    return 0;
}